### PR TITLE
Option to filter minimizer with bloomfilter

### PR DESCRIPTION
**Pull request by jowong4**
Deletion of `assert_good`  and the inclusion of `IOUtil.h` is due to compilation error since there is an assert_good function in the `btl_bloomfilter`.

Currently, the bloomfilter is a exclusion filter where if kmer `ATCG` is in the bloomfilter, it will not be a minimizer. I plan to add another flag `-mode` where we can exclude or include kmer in the bloomfilter

### CODE REVIEW

File: src/include/indexlr-minimize.h [hunk 1]
```diff
@@ -33,6 +35,22 @@ struct HashData
 	char strand;
 };
 
+struct find_HashData
```
Comment (schutzekatze) [inline src/include/indexlr-minimize.h]: Seems redundant, should be merged with HashData.

File: src/include/indexlr-minimize.h [hunk 2]
```diff
@@ -112,14 +130,51 @@ getMinimizers(const HashValues& hashes, const unsigned w)
 	return minimizers;
 }
 
-// Test the condition of a I/O stream.
-static inline void
-assert_good(const std::ios& stream, const std::string& path)
+static inline HashValues
+getMinimizers(const HashValues& hashes, const unsigned w, const BloomFilter& bloomFilter)
```
Comment (schutzekatze) [inline src/include/indexlr-minimize.h]: There seems to be a lot of code in common between this and the other `getMinimizers`. Can you separate that functionality in another function?

File: src/physlr-indexlr.cc [hunk 3]
```diff
@@ -26,9 +27,11 @@ minimizeReads(
     const size_t k,
     const size_t w,
     const size_t t,
+    const bool withBloomFilter,
     const bool withPositions,
     const bool withStrands,
-    const bool verbose)
+    const bool verbose,
+    BloomFilter& bloomFilter)
```
Comment (schutzekatze) [inline src/physlr-indexlr.cc]:
```suggestion
    const BloomFilter& bloomFilter)
```

File: src/physlr-indexlr.cc [hunk 4]
```diff
@@ -119,6 +135,15 @@ main(int argc, char* argv[])
 				          << ": Using more than 5 threads does not scale, reverting to 5.\n";
 			}
 			break;
+		case 'b':
+			withBloomFilter = true;
+			std::cerr << "loading bloomfilter from " << optarg << std::endl;
```
Comment (schutzekatze) [inline src/physlr-indexlr.cc]:
```suggestion
			std::cerr << "Loading bloomfilter from " << optarg << std::endl;
```

File: src/include/indexlr-minimize.h (hunk 1 above)
Comment (jowong4) [reply src/include/indexlr-minimize.h]: Removed the code as new implementation no longer requires it

File: src/include/indexlr-minimize.h (hunk 2 above)
Comment (jowong4) [reply src/include/indexlr-minimize.h]: Removed with new implementation

File: src/physlr-indexlr.cc (hunk 4 above)
Comment (jowong4) [reply src/physlr-indexlr.cc]: Fixed

File: src/physlr-indexlr.cc (hunk 3 above)
Comment (jowong4) [reply src/physlr-indexlr.cc]: fixed

**Comment by jowong4** [comment]
Any suggestions for alternative names of rBloomFilter and sBloomFilter. They stand for repetitive/ repeats Bloom filter and solid Bloom filter

File: data/Makefile [hunk 5]
```diff
@@ -642,11 +642,35 @@ minsize=2000
 
 # Index a FASTA file.
 %.k$k-w$w.physlr.tsv: %.fa
+ifeq ($(r),)
+ifeq ($(s),)
```
Comment (lcoombe) [inline data/Makefile]: Is there a default value for `r` and `s` in the Makefile? I'm thinking if at least there is a default setting we can add a comment to know what that variable means

File: src/physlr-indexlr.cc [hunk 6]
```diff
@@ -59,16 +77,18 @@ static void
...
-	             "  --strand   include minimizer strand in the output\n"
-	             "  -v         enable verbose output\n"
-	             "  -o FILE    write output to FILE, default is stdout\n"
-	             "  -t N       use N number of threads (default 1, max 5)\n"
-	             "  --help     display this help and exit\n"
-	             "  FILE       space separated list of FASTQ files\n";
+	          << "  -k K -w W [-r repeat_bf_path] [-s solid_bf_path] [-v] [-o FILE] FILE...\n\n"
+	             "  -k K        use K as k-mer size\n"
+	             "  -w W        use W as sliding-window size\n"
+	             "  -r repeat_bf_path  use a bloomfilter to filter out repetitive minimizers\n"
```
Comment (lcoombe) [inline src/physlr-indexlr.cc]:
```suggestion
	             "  -r repeat_bf_path  use a Bloom filter to filter out repetitive minimizers\n"
```

File: src/physlr-indexlr.cc [hunk 7]
```diff
@@ -59,16 +77,18 @@ static void
...
+	             "  -s solid_bf_path  use a bloomfilter to filter in solid minimizers\n"
```
Comment (lcoombe) [inline src/physlr-indexlr.cc]:
```suggestion
	             "  -s solid_bf_path  use a Bloom filter only select solid minimizers\n"
```
Just thinking 'filter in' doesn't sound right

File: src/vendor/btl_bloomfilter/pythonInterface/python_wrapper.readme [hunk 8]
```diff
@@ -0,0 +1,21 @@
+Python wrapper for BloomFilter.hpp
```
Comment (lcoombe) [inline src/vendor/btl_bloomfilter/pythonInterface/python_wrapper.readme]: Do we need to be including all of these interface files, etc. in this commit?

File: src/physlr-indexlr.cc [hunk 9]
```diff
@@ -4,9 +4,11 @@
 // Originally written for Physlr: (https://github.com/bcgsc/physlri
```
Comment (lcoombe) [inline src/physlr-indexlr.cc]:
```suggestion
// Originally written for Physlr: (https://github.com/bcgsc/physlr)
```

File: src/physlr-indexlr.cc [hunk 10]
```diff
@@ -146,6 +192,9 @@ main(int argc, char* argv[])
 	} else if (infiles.empty()) {
 		printErrorMsg(progname, "missing file operand");
 		failed = true;
+	} else if (withRepeat && withSolid) {
```
Comment (lcoombe) [inline src/physlr-indexlr.cc]: Can you clarify what type of error this statement is checking for?

Comment (jowong4) [reply src/physlr-indexlr.cc]: That was something I left in when I first wrote the code. Wasn't sure how to approach using both Bloom filter. I will remove this line

File: src/vendor/btl_bloomfilter/pythonInterface/python_wrapper.readme (hunk 8 above)
Comment (jowong4) [reply src/vendor/btl_bloomfilter/pythonInterface/python_wrapper.readme]: The intention was to just unzip the vendored repo and forget about it. Currently, abyss is like that too

File: data/Makefile (hunk 5 above)
Comment (jowong4) [reply data/Makefile]: I've added a comment and made the variables more verbose so people should understand the intention

File: src/vendor/btl_bloomfilter/pythonInterface/python_wrapper.readme (hunk 8 above)
Comment (lcoombe) [reply src/vendor/btl_bloomfilter/pythonInterface/python_wrapper.readme]: Ok makes sense!

**Pull request merged**